File: array/199.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> merge(vector<vector<int>>& intervals) {
    sort(intervals.begin(), intervals.end());
    vector<vector<int>> answer;
    vector<int> initialInterval = intervals[0];

    for (int i = 1; i < intervals.size(); ++i) {
        vector<int> temp = intervals[i];
        if (temp[0] == initialInterval[0]) {
            initialInterval[1] = temp[1];
        } else if (temp[0] <= initialInterval[1]) {
            initialInterval[1] = (temp[0] == initialInterval[1])  ? temp[1]
                                 : (initialInterval[1] > temp[1]) ? initialInterval[1]
                                                                  : temp[1];
        } else {
            answer.push_back(initialInterval);
            initialInterval = temp;
        }
    }

    answer.push_back(initialInterval);

    return answer;
}

vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
    bool inserted = false;
    vector<vector<int>> answer;

    if (intervals.size() == 0) {
        return {newInterval};
    }

    for (int i = 0; i < intervals.size(); ++i) {
        if (!inserted) {
            if (newInterval[0] < intervals[i][0]) {
                answer.push_back(newInterval);
                inserted = true;
            } else if (newInterval[0] == intervals[i][0] &&
                       newInterval[1] <= intervals[i][1]) {
                answer.push_back(newInterval);
                inserted = true;
            }
        }

        answer.push_back(intervals[i]);
    }

    if (!inserted) answer.push_back(newInterval);

    return merge(answer);
}

int main() {
    vector<vector<int>> intervals = {{1, 5}};
    vector<int> newInterval = {2, 7};

    for (const auto& v : insert(intervals, newInterval)) {
        for (const auto& e : v) {
            cout << e << " ";
        }

        cout << "\n";
    }

    

    return 0;
}

File: array/199.py
from typing import List


def merge(intervals: List[List[int]]) -> List[List[int]]:
    intervals.sort(key=lambda x: (x[0], x[1]))
    initialInterval = intervals[0]
    answer = []

    for index in range(1, len(intervals)):
        temp = intervals[index]

        if temp[0] == initialInterval[0]:
            initialInterval[1] = temp[1]
        elif temp[0] == initialInterval[1]:
            initialInterval[1] = temp[1]
        elif temp[0] < initialInterval[1]:
            if initialInterval[1] <= temp[1]:
                initialInterval[1] = temp[1]
        else:
            answer.append(initialInterval)
            initialInterval = temp

    answer.append(initialInterval)

    return answer


def insert(intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
    inserted = False
    answer = []

    if len(intervals) == 0:
        return [newInterval]

    for item in intervals:
        if inserted is False:
            if newInterval[0] < item[0]:
                answer.append(newInterval)
                inserted = True
            elif newInterval[0] == item[0] and newInterval[1] <= item[1]:
                answer.append(newInterval)
                inserted = True

        answer.append(item)

    if inserted is False:
        answer.append(newInterval)

    return merge(answer)


intervals = [[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]]
newInterval = [4, 8]

print(insert(intervals, newInterval))
